File: 515-find-largest-value-in-each-tree-row/find-largest-value-in-each-tree-row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int maxHeight(TreeNode* root)
    {
        if(root == NULL) return 0;
        return max(maxHeight(root->left) , maxHeight(root->right)) + 1;
    }

    void buildVector(TreeNode* root, vector<int>& vec, int level)
    {
        if(root == NULL) return;
        buildVector(root->left, vec, level+1);
        buildVector(root->right, vec, level+1);
        vec[level] = max(vec[level] , root->val);
    }

public:
    vector<int> largestValues(TreeNode* root) {
        int n = maxHeight(root);
        vector<int>vec(n,INT_MIN);
        buildVector(root,vec,0);
        return vec;
    }
};

File: 515-find-largest-value-in-each-tree-row/find-largest-value-in-each-tree-row.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 import java.util.Vector;
class Solution {
    public static int maxHeight(TreeNode root)
    {
        if(root == null) return 0;
        return Math.max(maxHeight(root.left) , maxHeight(root.right)) + 1;
    }

    public void buildVector(TreeNode root, Vector<Integer> vec, int level)
    {
        if(root == null) return;
        buildVector(root.left,vec, level+1);
        buildVector(root.right, vec, level+1);
        vec.set(level, Math.max(vec.get(level) , root.val));
    }

    public List<Integer> largestValues(TreeNode root) {
        int n = maxHeight(root);
        Vector<Integer>vec = new Vector<>();
        for(int i=0; i<n; i++)
        {
            vec.add(Integer.MIN_VALUE);
        }
        buildVector(root,vec,0);
        return vec;
    }
}